### PR TITLE
Bug fixes with kafka-go retry logic 

### PR DESCRIPTION
**Pull request by ShaneSaww**
1) Adding in more tests for kafka-go 
2)   Resetup the connection to the brokers when a message batch fails to produce. If you don't you will keep retrying the using the old (and possibly broken)  connection.

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -109,6 +112,32 @@ func (e Error) Temporary() bool {
 		e.Timeout()
 }
 
+// Retriable returns true if the error is one the "RETRIABLE" error codes from:
+// https://kafka.apache.org/protocol#protocol_error_codes
+func (e Error) Retriable() bool {
```
Comment (abuchanan-nr) [inline error.go]: Is this different than `Error.Temporary`?

Comment (ShaneSaww) [reply error.go]: @abuchanan-nr I'm using that now

File: writer.go [hunk 2]
```diff
@@ -727,53 +724,77 @@ func (w *writer) dial() (conn *Conn, err error) {
...
-			w.withErrorLogger(func(logger *log.Logger) {
-				logger.Printf("error dialing kafka brokers for topic %s (partition %d): %s", w.topic, w.partition, err)
-			})
-			for i, res := range resch {
-				res <- &writerError{msg: batch[i], err: err}
-			}
-			return
-		}
+func shouldRetry(err error, retries, attempts int) bool {
+	if retries == attempts {
```
Comment (colesnodgrass) [inline writer.go]: `if retries >= attempts` is a little safer in the event that `attempts` was accidentally incremented twice

**Review by colesnodgrass — APPROVED** [review]
One little check that could be improved, otherwise it looks good.

File: writer.go (hunk 2 above)
Comment (ShaneSaww) [reply writer.go]: yep, changed to that and updated some names to make it more clear

**Pull request merged**